### ISSUE TITLE
[Feature Request] Weight Normalization

### ISSUE THREAD

**Issue by Kaixhin** [pytorch/pytorch#1601]
[Layer normalization](https://arxiv.org/abs/1607.06450) seems to be pretty popular for RNNs nowadays, and it is worth having an implementation available. Several people seem to have already rolled their own (like @ajbrock), so this issue also aims to prevent several people working on this at the same time, as they can announce their intentions here.

The same applies for [weight normalization](https://arxiv.org/abs/1602.07868). This issue can be used to discuss both, but implementations could be done in separate PRs.

**Comment by ajbrock** [comment]
Good idea--does anyone have any ideas as to a faster way to do it then sending (0,1) transposed tensors through BatchNorm?

Relatedly, I've naively hacked out several weightnorm variants and normprop, which might both be worth having, though weightnorm would probably have to be a decorator of some sort rather than a separate layer.

**Comment by ethancaballero** [comment]
I tried applying this implementation of weightnorm https://github.com/ruotianluo/weightnorm-pytorch to a3c, but wasn't able to get it to work right.

**Comment by ajbrock** [comment]
It looks like that version includes the complicated data-dependent initialization, which I've also had trouble using even with the original Theano/Lasagne implementation provided by salimans. You can just use regular WeightNorm like this (verbose/naive version with the original formulation):

```
# 2D-conv with weightnorm
class Conv2d_WN(nn.Conv2d):
    def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=False):
        super(Conv2d,self).__init__(in_channels, out_channels, kernel_size, stride,
                 padding, dilation, groups, bias)
        self.g = Parameter(torch.ones(out_channels,1,1,1,1).expand_as(self.weight))
        
    def forward(self,x):
        wnorm = torch.sqrt((self.weight**2).sum(1).sum(2).sum(3)).expand_as(self.weight)
        return F.conv2d(x, self.g*self.weight/wnorm, bias=self.bias, stride=self.stride, padding=self.padding, dilation=self.dilation, groups=self.groups) 
```

There's a couple other variants lying around that in my experience work just as well and are even simpler.

**Comment by ethancaballero** [comment]
The easiest way to add weightnorm feature might be to create/handle weightnorm via the optimizer like in this implementation:
https://github.com/yaroslavvb/stuff/blob/master/keras_autoencoder/weightnorm.py#L75

**Comment by Jiaming-Liu** [comment]
This is also important for WGAN-GP.

**Comment by hyqneuron** [comment]
This pytorch implementation seems simple enough
https://gist.github.com/rtqichen/b22a9c6bfc4f36e605a7b3ac1ab4122f

**Comment by ajbrock** [comment]
That's exactly the kind of decorator I was hoping for! Though that version is implementing a simplified form, the original weightnorm is implemented per-output-channel, not over the whole tensor (though I've found that whole-tensor is way more stable, personally <_<)

**Comment by ethancaballero** [comment]
Has anyone gotten weightnorm (wn) to work well with RNNs for RL? Using wn with RNNs seems to perform the same or worse than without (wn) for me. I've tried applying to GRU/LSTM such that wn is used separately each component of weight matrices (6 components for GRU & 8 for LSTM) and tried applying wn to the whole (2) weight matrices.

**Comment by rtqichen** [comment]
Didn't realize about my gist being used! I actually had a much nicer version and I've updated the gist ( https://gist.github.com/rtqichen/b22a9c6bfc4f36e605a7b3ac1ab4122f ) with it. The code should be general enough for others to adapt it to their specific needs. (Though tbh I've never actually used it for anything.)

**Comment by soumith** [comment]
Weight normalization is now landed in master via #1945

**Comment by Kaixhin** [comment]
Pretty much all of the discussion here has been on WN so I'm marking this closed and moving LN to a separate feature request in order to make this tidier. Thanks everyone!
